Add section on expansion-time (early) name resolution

yaahc opened the pull request:
currently mostly a skeleton of a draft so we can collaboratively massage it into shape more easily before filling in with proper reference verbiage.

hoping to take a significant chunk out of https://github.com/rust-lang/reference/issues/568

yaahc committed:
Add section on expansion-time (early) name resolution

Co-authored-by: Vadim Petrochenkov <vadim.petrochenkov@gmail.com>
Co-authored-by: Eric Huss <eric@huss.org>
Co-authored-by: Tshepang Mbambo <hopsi@tuta.io>
Co-authored-by: Travis Cross <tc@traviscross.com>

yaahc reviewed (commented):
a fair number of comments are also contained inline in the documents

yaahc commented on src/items/use-declarations.md:
```diff
@@ -398,10 +398,18 @@ r[items.use.ambiguities]
+* except where shadowing is allowed
+r[names.resolution.early.imports.errors.ambiguity.globvsglob]
+* it is an error to name an item through ambiguous use declarations
+    * two globs imports which both have an item matching that name where the items are different
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=11d42f996c2a1b12c922973aed10229d

yaahc commented on src/items/use-declarations.md:
```diff
@@ -398,10 +398,18 @@ r[items.use.ambiguities]
+* it is an error to name an item through ambiguous use declarations
+    * two globs imports which both have an item matching that name where the items are different
+        * this is still an error even if there is a third non glob binding resolution to an item with the same name
+* it is not an error to have two glob imports which include items which would be ambiguous so long as you do not name one of those items through the ambiguous glob imports
```
documented

petrochenkov commented on src/items/use-declarations.md:
```diff
@@ -398,10 +398,18 @@ r[items.use.ambiguities]
+r[names.resolution.early.imports.errors.ambiguity.globvsglob]
+* it is an error to name an item through ambiguous use declarations
+    * two globs imports which both have an item matching that name where the items are different
+        * this is still an error even if there is a third non glob binding resolution to an item with the same name
```
>I expected this to be dependent upon ordering.

Resolution results are supposed to never depend on the item ordering (besides `macro_rules` with their "textual scope"), otherwise we won't be able to have things like [parallel import resolution](https://github.com/rust-lang/rust/pull/145108) or import reordering in rustfmt.
In practice there may be dependencies in corner cases, but every such case is a bug.

petrochenkov commented on src/items/use-declarations.md:
```diff
@@ -442,6 +449,111 @@ fn main() {
+      sure if this actually exists or is just proposed, TODO investigate)
+    * TODO example? This ones harder to do concisely afaik
+r[names.resolution.early.imports.errors.ambiguity.derivehelper]
+* derive helpers used before their associated derive may not shadow other attributes or other derive helpers that are otherwise in scope after their derive
```
This is something that exists under a deprecation lint in the final stage of its lifecycle (https://github.com/rust-lang/rust/issues/79202), and hopefully will be removed next year.
Not sure if things like this are supposed to be documented in the reference.

yaahc replied:
ive gone ahead and moved this to be in the proc macro helper attribute section and setup a first pass attempt at documenting the intended behavior as I understand it, the admonition about deprecation, and an example.

yaahc commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,106 @@
+
+r[names.resolution.early.imports.shadowing]
+
+The following is a list of situations where shadowing of use declarations is permitted:
```
I think the original text is correct.

Use glob shadowing: https://doc.rust-lang.org/nightly/reference/items/use-declarations.html#r-items.use.glob.shadowing
macro textual scope shadowing: https://doc.rust-lang.org/nightly/reference/macros-by-example.html#r-macro.decl.scope.textual.shadow

In my mind the "not permitted" set is the list of ambiguity errors below

yaahc commented on src/items/use-declarations.md:
```diff
@@ -442,6 +449,111 @@ fn main() {
+    * [macro.decl.scope.path.ambiguity]
+r[names.resolution.early.imports.errors.ambiguity.globvsouter]
+* it is an error to shadow an outer name binding with a glob import
+    * This seems to only apply to early resolution (duh, I documented this as part of an early resolution codepath)
```
I don't mind moving this but the reason I put them alongside their items was because that seemed most consistent with how we documented other parts of name resolution, specifically how the names that are introduced by each item are documented next to their items and then theres an aggregated list linking to all those sections in the namespaces chapter.

My plan was to put ambiguity and shadowing sections alongside their items and aggregate links to all those sections in name-resolution.md.

You can see where I discussed this plan with eric starting here: [#t-lang-docs/reference > name resolution @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/520709-t-lang-docs.2Freference/topic/name.20resolution/near/544623109)

yaahc commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,106 @@
+* Names are resolved at three different stages of compilation.
+* [Macros] and [use declarations] are resolved during macro expansion.
+    * This stage of resolution is known as "Early Resolution".
+* Associated consts and functions, methods, and enum variants are resolved during type checking.
```
ACK

I went ahead and vaguely updated the associated section but I bet its still wrong. Don't worry about it too much until after I've had time to go thru and properly explore type-relative resolution and understand the difference between these cases.

ehuss commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,114 @@
+r[names.resolution]
```
Let's just delete the duplicates: https://github.com/rust-lang/reference/pull/2060

yaahc commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,416 @@
+and expanding macro invocations, then repeating until there are no further
+macros invocations to resolve. Once this process is completed all the imports
+are resolved again to ensure that the macro expansion process did not introduce
+any new ambiguious imports.
```
do we want to talk about this? feels like an implementation detail but
also really helps to understand certain kinds of ambiguity errors that users
can run into.

yaahc commented on src/procedural-macros.md:
```diff
@@ -440,3 +444,4 @@ their equivalent `#[doc = r"str"]` attributes when passed to macros.
 [type expressions]: types.md#type-expressions
 [type]: types.md
 [union]: items/unions.md
+[name resolution ambiguities]: ../names/name-resolution.html#r-names.resolution.expansion.imports.errors.ambiguity
```
more precise link

petrochenkov commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,351 @@
+Some situations are an error when there is an ambiguity as to which name a `use` declaration refers. This happens when there are two name candidates that do not resolve to the same entity where neither import is [permitted] to shadow the other.
+
+r[names.resolution.expansion.imports.ambiguity.globvsglob]
+Names may not be resolved through ambiguous glob statements. Glob imports are allowed to import conflicting names in the same namespace as long as the name is not used. Conflicting names from ambiguous glob statements may still be shadowed and used without producing an error. The errors occur at time of use, not time of import.
```
```
Conflicting names from ambiguous glob statements may still be ... used
```
Weird wording, they both can and cannot be used.
I guess the intent was to say that the ambiguous globs can be shadowed, and the thing that shadows then can then be used?

petrochenkov commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,416 @@
+* derive helpers used before their associated derive may not shadow other attributes or other derive helpers that are otherwise in scope after their derive
+    * TODO example? This ones harder to do concisely afaik
+
+Helper attributes may not be used before the macro that introduces them.
```
Would be nice to add a test for this to rust-lang/rust, if there's none currently.

yaahc replied:
I was setting up a test for this and I think I found a mistake I made in interpreting the logic in `resolve_ident_in_scope_set`. Specifically I said 

```rust
                                } else if innermost_res == derive_helper_compat
                                    || res == derive_helper_compat && innermost_res != derive_helper
                                {
                                    Some(AmbiguityKind::DeriveHelper)
```

means

- if a derive helper which is used before it's derive has the same name as an outer item
- or if a derive helper has the same name as an outer derive helper which is used before it's derive

The first bullet is true afaict, being in this codepath implies theres a res and innermost_res, so having the innermost be a derive_helper_compat means that we have a second candidate we found later in the same sub-namespace.

The second bullet is incorrect. It seems like I interpreted the `!=` as an `==`. This is just saying that the second resolution is a compat resolution, and the first resolution is not a helper, which means it could only be another helper compat due to visitation order. Given the lack of identity on helpers, I'm not sure how to trigger this second half of the condition.

Here's the test I setup:

```rust
#[empty_helper]
#[derive(Empty)]
#[empty_helper]
#[derive(Empty2)]
struct S;
```

Both derives introduce the same `#[empty_helper]` attribute. My understanding was that when visiting scopes we'd see the introduction of `#[empty_helper]` from both derives, which would populate res and innermost_res in a way that should trigger the DeriveHelper ambiguity error. Instead, the only error this produces is the future incompat warning for using a helper before it's derive which is converted into an error by `deny(future_incompat)`.

I'm going to try digging into this a bit deeper to see if I can pinpoint exactly when this subbranch gets triggered or whether I'm correct in thinking this is inert.

yaahc commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,351 @@
 
+r[names.resolution.intro]
+
+_Name resolution_ is the process of tying paths and other identifiers to the declarations of those entities. Names are segregated into different [namespaces], allowing entities in different namespaces to share the same name without conflict. Each name is valid within a [scope], or a region of source text where that name may be referenced. Access to certain names may be restricted based on their [ ...
```
lifetimes / type parameters come to mind as an example which I wouldn't consider paths that need their names resolved.

yaahc commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,351 @@
+## Ambiguities
+
+r[names.resolution.expansion.imports.ambiguity.intro]
+Some situations are an error when there is an ambiguity as to which name a `use` declaration refers. This happens when there are two name candidates that do not resolve to the same entity where neither import is [permitted] to shadow the other.
```
you are correct. I went ahead and added the extra cases and improved the examples for globvsouter to better demonstrate the variety of cases

yaahc commented:
@rustbot label -S-waiting-on-author

petrochenkov commented on src/names/name-resolution.md:
````diff
@@ -1,4 +1,356 @@
+}
+```
+
+Multiple glob imports are allowed to import the same name, and that name is allowed to be used, if the imports are of the same item (following re-exports). The visibility of the name is the maximum visibility of the imports. For example:
````
>The visibility of the name is the maximum visibility of the imports.

The rules here will need to be restricted to address the so called "`RustEmbed` issue" (https://github.com/rust-lang/rust/pull/147196).
https://github.com/rust-lang/rust/pull/149145 starts working on the solution.

Multiple competing bindings with different visibilities are typically not a problem because visibilities are checked (or at least re-checked) after expansion (during macro/import validation pass in early resolution, or later).
The problem is that visibility *can* be consumed early without later validation, and then propagated uncontrollably, if the ambiguous binding is reexported further.
```rust
// Pseudocode

vis1 use m1::*; // imports `Item`
vis2 use m2::*; // imports same `Item`

vis3 use self::Item;
// How far will `Item` be reexported, to `vis1`, to `vis2` or to `vis3`?
// The answer is unpredictable and depends on the resolution/expansion order.
// This is not an issue if `min(vis1, vis3) == min(vis2, vis3)`.
```

ehuss commented on src/items/use-declarations.md:
```diff
@@ -460,3 +412,4 @@ fn main() {
 [tool attributes]: ../attributes.md#tool-attributes
 [type alias]: type-aliases.md
 [type namespace]: ../names/namespaces.md
+[name resolution ambiguities]: ../names/name-resolution.html#r-names.resolution.expansion.imports.ambiguity
```
These kinds of links should just be the rule name.


```suggestion
[name resolution ambiguities]: names.resolution.expansion.imports.ambiguity
```

ehuss commented on src/macros-by-example.md:
````diff
@@ -326,6 +326,77 @@ fn foo() {
 ```
 
+r[macro.decl.scope.textual.shadow.path-based]
+Textual scope name bindings for macros may shadow path-based scope bindings to macros.
````
The use of the word "may" here feels like it is weakening the statement, or making it sound like it is optional. I'm not sure if that was intentional or if that word is needed here? That is, could it just be dropped?


```suggestion
Textual scope name bindings for macros shadow path-based scope bindings to macros.
```

tshepang commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,348 @@
+Helper attributes may not be used before the macro that introduces them.
+
+> [!NOTE]
+> rustc currently allows derive helpers to be used before their attribute macro introduces them into scope so long as they do not shadow any other attributes or derive helpers that are otherwise correctly in scope. This behavior deprecated and slated for removal.
```
```suggestion
> rustc currently allows derive helpers to be used before their attribute macro introduces them into scope so long as they do not shadow any other attributes or derive helpers that are otherwise correctly in scope. This behavior is deprecated and slated for removal.
```

ehuss commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,348 @@
+The following is a list of situations where shadowing of use declarations is permitted:
+
+* [use glob shadowing]
+* [macro textual scope shadowing]
```
Lists should usually start with a capital letter. Also, I realize we are not very consistent with this, but I think in this case the `use` would be inline code to differentiate it from the English word "use".

```suggestion
* [`use` glob shadowing]
* [Macro textual scope shadowing]
```

ehuss commented:
Something I'm missing is a discussion about blocks, and how they can introduce essentially an anonymous module (`ModuleKind::Block`). That's important when trying to understand the module tree and how some resolutions and ambiguities resolve differently. I was looking at this example:

```rust
mod bar {
    pub struct Name;
}

mod baz {
    pub struct Name;
}

use baz::Name;

pub fn foo() {
    use bar::*;
    Name; // resolves to bar::Name
}
```

I think the reason it resolves to `bar::Name` is because it is nested inside an anonymous module. If we were to rewrite this to:

```rust
mod bar {
    pub struct Name;
}

mod baz {
    pub struct Name;
}

use baz::Name;
use bar::*;

pub fn foo() {
    Name; // resolves to baz::Name
}
```

If we were to ignore blocks, and pretend they don't exist from a module-hierarchy perspective, I would expect both these examples to be the same. I'm not sure if I'm understanding this correctly, since resolution within a block still "sees" everything outside of the block (inside it's module).

ehuss commented on src/procedural-macros.md:
````diff
@@ -240,6 +240,26 @@ A helper attribute for a derive macro is declared by adding its identifier to th
 > ```
 
+r[macro.proc.derive.attributes.scope]
+Helper attributes are in scope after the macro that introduces them.
````
Would it be possible to be more specific about what "after" means here? There's some start point and some end point, correct?

That is, there are different interpretations of what "after" means:
- Anything after that point to the end of the file.
- Only *inside* the attached item.
- Inside the attached item *and* inside the output of the derive.
- Inside the attached item *and* any outer attribute *after* the `derive`.

The last one is my understanding of what it covers. I'm not sure what the best way to phrase that is.

ehuss commented on src/procedural-macros.md:
```diff
@@ -240,6 +240,26 @@ A helper attribute for a derive macro is declared by adding its identifier to th
+Helper attributes are in scope after the macro that introduces them.
+
+> [!NOTE]
+> rustc currently allows derive helpers to be used before the macro that
```
Unwrap the blockquote.

ehuss commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,384 @@
+
+> [!NOTE]
+>
+> The compiler will attempt to resolve derive helpers that are used before
```
Unwrap.

(I realize blockquotes don't look terribly great in source. I'm wondering if we should just choose a different syntax...)

ehuss commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,384 @@
+> Starting in edition 2018 the `#[macro_use]` prelude is not visited when `#[no_implicit_prelude]` is present.
+
+r[names.resolution.expansion.macros.derivehelpers]
+not visited when resolving derive macros in the parent scope (starting scope)
```
Seems like an incomplete rule?

ehuss commented on src/names/name-resolution.md:
````diff
@@ -1,4 +1,384 @@
+```
+
+> [!NOTE]
+> These ambiguity errors are specific to imports, even though they are only observed when those imports are used, having multiple candidates available for a given name during later stages of resolution is not considered an error, so long as none of the imports themselves are ambiguous, there will always be a single unambiguous closest resolution during later stages.
````
Can this be split up into a few sentences?  (and the "during later stages" seems duplicated.)

Maybe something like:


```suggestion
> These ambiguity errors are specific to imports, even though they are only observed when those imports are used. Having multiple candidates available for a given name during later stages of resolution is not considered an error, so long as none of the imports themselves are ambiguous, there will always be a single unambiguous closest resolution.
```

yaahc commented on src/names/name-resolution.md:
```diff
@@ -1,4 +1,420 @@
+* [Path-based scope macros]
+* [`macro_use` prelude]
+* [Standard library prelude]
+* [Builtin attributes]
```
should this clarify that each of these bullets represent one or more scopes or is this intuitively obvious?

ehuss commented on src/names/name-resolution.md:
````diff
@@ -1,4 +1,553 @@
+ambig!(); // ERROR: `ambig` is ambiguous.
+```
+
+This also applies to imports, so long as the innermost candidate for the invocation of name is from within a macro expansion.
````
We were puzzling over this sentence and were wondering if you could clarify something.

(The phrase "the invocation of name" seemed out of place here, and we are assuming this should be "the name", since there isn't an invocation here.)

This ambiguities section starts off talking about an ambiguity involving macro invocations and expanded macros. But then this sentence seems to say, "oh and the ambiguity applies to imports, too.". Am I understanding it correctly that the rule here applies to macro invocations **and** imports? Is there a reason the start of the rule leaves out the "and imports" part?

yaahc replied:
The "invocation" language came up because of [this function](https://github.com/rust-lang/rust/blob/main/compiler/rustc_resolve/src/lib.rs#L1024) and how it refers to the site the name being resolved came from as an invocation regardless of whether it was an import or a macro. I wasn't really sure if this was the more proper term or not, elsewhere in the reference I've used language like "name is used" to the same effect. 

I just reread the beginning of this ambiguity section and at least the way I read and meant it it's not talking about this as only applying to macros. Whether or not the invocation is a macro or a use decl there have to be macro expansions involved in order for this lint to trigger. I did my best to make the language generic in the beginning and talk about candidates without specifically saying imports or macro definitions as the sources of those candidate bindings.

traviscross committed:
Make final editorial revisions on name resolution

Many earlier passes of revisions have been squashed into the chapter
itself.  These edits are the final pass.

traviscross committed:
Remove `derive-helpers` rule with unclear meaning

We're not certain what this rule means; we've run various tests and
could not confirm our interpretation of the rule.  E.g. (thanks to
ehuss for this):

```rust
use pm::{WithHelperAttr, helper};
#[derive(WithHelperAttr)]
#[helper]
#[derive(helper)] // ERROR: expected derive macro, found derive helper attribute `helper`
struct S;
```

Interestingly, this works and calls the attribute macro:

```rust
use pm::{WithHelperAttr, helper};
#[derive(WithHelperAttr, helper)]
#[helper]
struct S;
```

So that we can merge the rest, let's remove this rule for now and add
it back in future work.
